presto-main/src/main/java/com/facebook/presto/operator/scalar/CharacterStringCasts.java

```diff
@@ -69,31 +71,30 @@ public static Slice varcharToCharCast(@LiteralParameter("y") Long y, @SqlType("v
     @LiteralParameters({"x", "y"})
     public static Slice charToVarcharCast(@LiteralParameter("y") Long y, @SqlType("char(x)") Slice slice)
     {
-        int textLength = countCodePoints(slice);
-        int resultLength = y.intValue();
+        return padSpacesAndTruncateToLength(slice, y.intValue());
+    }
 
-        // if our target length is the same as our string then return our string
-        if (textLength == resultLength) {
-            return slice;
-        }
+    @ScalarOperator(OperatorType.SATURATED_FLOOR_CAST)
```

Comment:
Saturated foor casts are internal operators (hidden) that are used by `DomainTranslator` in order to compute tuple domain predicate for connector.

They help to deal with predicates of different (but coercable) types, e.g:
`col_bigint = DOUBLE 0.01`
or
`col_char = 'foo'`
In such cases you want tuple domain of `BIGINT` or `CHAR` type, but the literal has wider type. Floor casts are used to rewrite such predicates so that both operands have the same type (`BIGINT` or `CHAR` in this case). The idea is that floor cast returns value in target type (e.g. type with less precision) that is smaller than source literal value, but as large as possible in target type. Based on that it is possible to rewrite multi-type predicate to a corresponding predicate with one, less precise type. The rewritten predicate is used only to compute tuple domain used in connectors to read data.

See `DomainTranslator#rewriteComparisonExpression`


Strictly floor cast should return value smaller than the argument, but as large as possible in target type. However, `DomainTranslator` algorithms works with any value that is just smaller than the argument (not necessarily the largest one). In such case, the tuple domain predicate used in connectors is be more selective, but still correct. This is the case I simplified floor cast from `VARCHAR` to `CHAR` so that it just returns reasonable smaller value.

cc: @arhimondr